overlord/servicestate/quota_control_test.go

```diff
@@ -134,14 +153,103 @@ func checkSliceState(c *C, sliceName string, sliceMem quantity.Size) {
 	}
 }
 
-func (s *quotaControlSuite) TestCreateQuota(c *C) {
-	r := s.mockSystemctlCalls(c, []expectedSystemctl{
+func systemctlCallsForSliceStart(name string) []expectedSystemctl {
+	slice := "snap." + name + ".slice"
+	return []expectedSystemctl{
+		{expArgs: []string{"start", slice}},
+	}
+}
+
+func systemctlCallsForSliceStop(name string) []expectedSystemctl {
+	slice := "snap." + name + ".slice"
+	return []expectedSystemctl{
+		{expArgs: []string{"stop", slice}},
 		{
-			// called for new slice unit written by CreateQuota after we create
-			// the snap in state
-			expArgs: []string{"daemon-reload"},
+			expArgs: []string{"show", "--property=ActiveState", slice},
+			output:  "ActiveState=inactive",
+		},
+	}
+}
+
+func systemctlCallsForServiceRestart(name string) []expectedSystemctl {
+	svc := "snap." + name + ".svc1.service"
+	return []expectedSystemctl{
+		{
+			expArgs: []string{"is-enabled", svc},
+			output:  "enabled",
+		},
+		{expArgs: []string{"stop", svc}},
+		{
+			expArgs: []string{"show", "--property=ActiveState", svc},
+			output:  "ActiveState=inactive",
+		},
+		{expArgs: []string{"start", svc}},
+	}
+}
+
+func systemctlCallsForCreateQuota(groupName, snapName string) []expectedSystemctl {
+	return join(
+		[]expectedSystemctl{{expArgs: []string{"daemon-reload"}}},
+		systemctlCallsForSliceStart(groupName),
+		systemctlCallsForServiceRestart(snapName),
+	)
+}
+
+func systemctlCallsVersion(version int) []expectedSystemctl {
+	return []expectedSystemctl{
+		{
+			expArgs: []string{"--version"},
+			output:  fmt.Sprintf("systemd %d\n+FOO +BAR\n", version),
+		},
+	}
+}
+
+func join(calls ...[]expectedSystemctl) []expectedSystemctl {
+	fullCall := []expectedSystemctl{}
+	for _, call := range calls {
+		fullCall = append(fullCall, call...)
+	}
+
+	return fullCall
+}
+
+func (s *quotaControlSuite) TestCreateQuotaPreseeding(c *C) {
```

Comment:
Silly question again - under what situation would quota groups be created when pre-seeding? Currently we only support creating them via the API, yes? So if this is for forward compatibility I wonder if it's worth it, because AIUI we would have to create the slice units and put them on disk too. So instead of anticipating pre-seed now with quota groups we could simply "panic" when quota groups are created in preseed mode. Once we reach the point when quota-groups get added to e.g. gadget.yaml or similar so that they are no longer an API only feature we would do the work to fully support them.

Or is this too simplistic? I.e. I'm missing something (obvious) ?

Reply:
I guess I would prefer us to just leave the code as-is here since I've already written it and as the unit test demonstrates it does work and it's not terribly complicated nor semantically wrong - the code as-is _does_ write the slice units on disk when preseeding is on, we just don't have a way to trigger these methods during preseeding. When we want/need to enable this, we would just need a way to call these servicestate methods during preseeding, which as you say would be very similar to the gadget case.

The only exception to this would be removing quotas that Samuele mentioned below, that doesn't make sense and should be modified to error out on preseeding. 

If you really want me to rip it out, I can, it's always just a bit annoying to have to delete code that isn't obviously wrong nor complicated.